pkg/hostfimsensor/v1/filetree/filetree.go

```diff
@@ -0,0 +1,249 @@
+func NewFileNode(name, path string, isDir bool) *FileNode {
+	return &FileNode{
+		Name:     name,
+		Path:     path,
+		IsDir:    isDir,
+		Children: make(map[string]*FileNode),
+	}
+}
+
+// AddChild adds a child node to this node
+func (n *FileNode) AddChild(child *FileNode) {
+	n.mu.Lock()
+	defer n.mu.Unlock()
+	
+	child.Parent = n
+	n.Children[child.Name] = child
+}
+
+// GetChild retrieves a child node by name
+func (n *FileNode) GetChild(name string) *FileNode {
+	n.mu.RLock()
+	defer n.mu.RUnlock()
+	
+	return n.Children[name]
+}
+
+// HasChild checks if a child with the given name exists
+func (n *FileNode) HasChild(name string) bool {
+	n.mu.RLock()
+	defer n.mu.RUnlock()
+	
+	_, exists := n.Children[name]
+	return exists
+}
+
+// RemoveChild removes a child node
+func (n *FileNode) RemoveChild(name string) {
+	n.mu.Lock()
+	defer n.mu.Unlock()
+	
+	if child, exists := n.Children[name]; exists {
+		child.Parent = nil
+		delete(n.Children, name)
+	}
+}
+
+// GetChildren returns a copy of all children
+func (n *FileNode) GetChildren() map[string]*FileNode {
+	n.mu.RLock()
+	defer n.mu.RUnlock()
+	
+	children := make(map[string]*FileNode)
+	for name, child := range n.Children {
+		children[name] = child
+	}
+	return children
+}
+
+// SetMetadata sets the file metadata from os.FileInfo and syscall.Stat_t
+func (n *FileNode) SetMetadata(info os.FileInfo, stat *syscall.Stat_t) {
+	n.mu.Lock()
+	defer n.mu.Unlock()
+	
+	n.Size = info.Size()
+	n.Mode = uint32(info.Mode())
+	n.Mtime = info.ModTime()
+	n.Ctime = info.ModTime() // Use mtime as fallback for ctime
+	
+	if stat != nil {
+		n.Inode = stat.Ino
+		n.Device = uint64(stat.Dev)
+		n.Uid = stat.Uid
+		n.Gid = stat.Gid
+	}
+}
+
+// GetNodeCount returns the total number of nodes in the tree
+func (t *FileTree) GetNodeCount() int {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	return t.NodeCount
+}
+
+// IncrementNodeCount increments the node count
+func (t *FileTree) IncrementNodeCount() {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	t.NodeCount++
+}
+
+// SetRoot sets the root node of the tree
+func (t *FileTree) SetRoot(root *FileNode) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	t.Root = root
+	if root != nil {
+		t.NodeCount = 1 // Root node counts as 1
+	}
+}
+
+// GetRoot returns the root node
+func (t *FileTree) GetRoot() *FileNode {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	return t.Root
+}
+
+// FindNode finds a node by path
+func (t *FileTree) FindNode(path string) *FileNode {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	
+	if t.Root == nil {
+		return nil
+	}
+	
+	if path == "/" || path == "" {
+		return t.Root
+	}
+	
+	// Split path and traverse
+	parts := splitPath(path)
+	current := t.Root
+	
+	for _, part := range parts {
+		if current == nil || !current.IsDir {
+			return nil
+		}
+		current = current.GetChild(part)
+	}
+	
+	return current
+}
+
+// splitPath splits a path into components, handling both absolute and relative paths
+func splitPath(path string) []string {
```

Comment:
I'd be careful here between the split and the join elsewhere - we might be doing lots of allocations for nothing, wdyt? I don't know how deep the filesystemps could be on a node...

Reply:
You're right! I am changing the implementation to deconstruct the path once and use a slice all the way so no need to join/split every time